### ISSUE TITLE
Use memoryfs to speed up?

### ISSUE THREAD

**Issue by davelab6** [googlefonts/fontmake#798]
@twardoch suggested on a call today that fontmake could as an option use https://docs.pyfilesystem.org/en/latest/reference/memoryfs.html to make a ram disk and write its temporary process files to there instead of a real (slow) disk.

**Comment by alerque** [comment]
I tried out a rough approximation of this this out on a project ([Raleway](https://github.com/theleagueof/raleway)) using [`fontship`](https://github.com/theleagueof/fontship). Because *everything* is done in a scratch directory anyway it is pretty easy to provide it with a RAM disk. The "approximation" factor lines in the fact that everything in the entire chain including not just `fontmake` but other steps in the process are included in the test and are all configured to use the same scratch disk (everything between input sources and final font files).

* Disk: 53.9s
* RAM: 54.2s

I reran a couple times and got jitter in the 1s range, so the result is a draw.

I believe the (lack of) difference is that the Disk in question is a (budget priced mid grade) NVe SSD and hence relatively quick to start with. It would be interesting to try the same thing on spinning rust disks with small caches, but I suspect the answer will be there is no appreciable win here. One of the reasons I think even spinning Rust disks won't show a huge speed decrease is that the file sizes in question are all massively under the size of most disk read/write caches. Even if the actual magnetic medium doesn't get updated until long after, the write / read cycle will have completed from the disk's fast cache memory anyway.

At the end of the day I believe any slim benefit to be squeezed out would be negated by the time it took Python to setup and tear down the ramdisk. A build orchestration system (Fontship or gftools or whatever) that could better manage a long-running RAM disk for use in repeated runs and other tooling as well would make much more sense, but so far my testing (both now and a while back) suggests there isn't a lot to be gained.

On the other hand swapping out bits of fontmake a piece at a time for Rust equivalents could possibly change things by an order of magnitude.

**Comment by anthrotype** [comment]
fs.MemoryFS is not a ram disk, it's just an implementation of the fs interface that doesn't actually read/write anything to the filesystem proper, but in which "files" are actually in-memory `io.BytesIO` file-like objects
https://github.com/PyFilesystem/pyfilesystem2/blob/master/fs/memoryfs.py

**Comment by davelab6** [comment]
I'm sorry but I'm unsure why that isn't a ramdisk. Isn't that what a ram disk does?

**Comment by felipesanches** [comment]
> I'm sorry but I'm unsure why that isn't a ramdisk. Isn't that what a ram disk does?

Operating systems can provide a "ram disk" at kernel level, meaning a storage device that uses RAM and can be used regularly (format a filesystem, save files, etc).

That's not the same as implementing a python class with the same interface as file-like objects but using `io.BytesIO`. In the end, the data also ends up in RAM, instead of a hard-drive. But it uses a different mechanism to to so. I am not sure, but there may be differences in performance for the two approaches.

Also, operating systems tend to be smart when performing disk I/O operations. For instance, I use a USB-floppy drive here sometimes, and whenever I make modifications to a 1MB file on floppy I can observe that the actual operation of flushing the data to the disk is delayed, perhaps in the hope that additional changes will still be made in the short term. Un-mounting the disk, though, immediately flushes the data to be committed to the physical medium.

I guess what this means is that at least some optimizations are probably already done for us at the kernel level when dealing with slow I/O devices.

**Comment by alerque** [comment]
I think @felipesanches is right here. Wherever they are, some sort of optimizations seem to be already making up for any holdups that would be caused by slow I/O. This is not to say that doing such expensive I/O operations in the first place is a good idea, just that trying to deal with them by optimizing them a different way than is happening automatically won't like lead to much gain.

For context the reason I looked into this in the first place is that I have 2 large build orchestration systems: one builds books (CaSILE) and the other builds fonts (Fontship). Both have the same overall architecture (and indeed share most of their Rust code in common). Both create a a scratch space, run a whole bunch of shell commands to get artifacts through intermediate forms from one build stage to another, then write the final form out to the original directory. The difference is CaSILE projects involve many more stages (and hence disk round trips) and much larger data sets (250MB–5Gig) per build and some stages can have very large intermediary files, while Fontship projects tend to have lots of small files (hello UFOs) but the sum total of all intermediate files written is usually 20-100MB) and usually any one intermediate stage written to disk is no more than a few MB.

When I experimented with putting CaSILE builds on a RAM disk the performance gains were significant, but it required a system with quite a bit of spare RAM to work well. With that in mind I decided to try it with Fontship and ... got what you see above: no measurable wins.

----

Instead of trying to slip a faster file system in the middle, I think it would be worth looking at why you're using intermediate files at all. Even writing them to a file-like-API is just a round trip that apparently doesn't need to be happening. For speed you'll want an in memory hand-off of the original or some serialized structure to the next function(s).
